pkg/operator/ceph/object/user/controller.go

```diff
@@ -278,41 +284,105 @@ func (r *ReconcileObjectStoreUser) reconcile(request reconcile.Request) (reconci
 	return reconcile.Result{}, *cephObjectStoreUser, nil
 }
 
-func (r *ReconcileObjectStoreUser) reconcileCephUser(cephObjectStoreUser *cephv1.CephObjectStoreUser) (reconcile.Result, error) {
-	err := r.createOrUpdateCephUser(cephObjectStoreUser)
+func (r *ReconcileObjectStoreUser) reconcileCephUser(cephObjectStoreUser *cephv1.CephObjectStoreUser) (reconcile.Result, bool, error) {
+	userKeysSource, err := r.createOrUpdateCephUser(cephObjectStoreUser)
 	if err != nil {
-		return reconcile.Result{}, errors.Wrapf(err, "failed to create/update object store user %q", cephObjectStoreUser.Name)
+		return reconcile.Result{}, userKeysSource, errors.Wrapf(err, "failed to create/update object store user %q", cephObjectStoreUser.Name)
 	}
 
-	return reconcile.Result{}, nil
+	return reconcile.Result{}, userKeysSource, nil
+}
+
+func validateSecretKeys(accessKey, secretKey string, secretKeys []admin.UserKeySpec) ([]admin.UserKeySpec, error) {
+	// if len is 0 just update the secretKeys array with the accessKey and secretKey key
+	if len(secretKeys) == 0 {
+		secretKeys = []admin.UserKeySpec{
+			{AccessKey: accessKey,
+				SecretKey: secretKey}}
+		return secretKeys, nil
+	}
+	// match if the secretKeys array contains the accessKey and secretKey key
+	for i := range secretKeys {
+		if secretKeys[i].AccessKey == accessKey && secretKeys[0].SecretKey == secretKey {
+			return secretKeys, nil
+		}
+	}
+
+	return secretKeys, errors.Errorf("secret keys data is invalid, please update the secret with cvalid format")
+}
+
+func forceUpdateObjectUserSecret(annotations map[string]string) bool {
+	if value, found := annotations[updateObjectUserSecretAnnotation]; found {
+		if strings.EqualFold(value, sourceOfTruthSecret) {
+			return true
+		}
+	}
+	return false
+}
+
+func getSecretKeysJsonStruct(secretKeys []byte) ([]admin.UserKeySpec, error) {
+	var userKeys []admin.UserKeySpec
+	err := json.Unmarshal(secretKeys, &userKeys)
+	if err != nil {
+		return userKeys, errors.Wrapf(err, "unable to unmarshal secretKeys from the object secret")
+	}
+	return userKeys, nil
 }
 
-func (r *ReconcileObjectStoreUser) createOrUpdateCephUser(u *cephv1.CephObjectStoreUser) error {
+func (r *ReconcileObjectStoreUser) createOrUpdateCephUser(u *cephv1.CephObjectStoreUser) (bool, error) {
```

Comment:
Instead of creating a separate credential struct, I am using the `UserKeySpec` struct, which is defined in goceph.
Benefits:
1) Easy to marshal
2) All the rgw go ceph API calls need this,  if we create something specific we then again need to update the `UserKeySpec`
3) In future if anything more required for the secret we can just update the credentials, and it will be auto handled
```
type UserKeySpec struct {
	User      string `json:"user"`
	AccessKey string `json:"access_key" url:"access-key"`
	SecretKey string `json:"secret_key" url:"secret-key"`
	// Request fields
	UID         string `url:"uid"`     // The user ID to receive the new key
	SubUser     string `url:"subuser"` // The subuser ID to receive the new key
	KeyType     string `url:"key-type"`
	GenerateKey *bool  `url:"generate-key"` // Generate a new key pair and add to the existing keyring
}
```

Reply:
We don't want to marshal all of this stuff into the `Credentials` info for users, who will find it unnecessary and confusing. Generally, things work best when the spec that users see via the API is a subset of the spec used internally. But APIs become "leaky abstractions" when the internal structs are exposed to users in API surfaces. These leaky abstractions are usually complicated for users to understand, and because they have fields that users aren't supposed to be allowed to interact with, their behavior is often undefined, or worse, broken.

We should only put stuff into Credentials that we want users to be able to modify. That means we should define a separate spec for what users see. 

We can retain code reuse by using golang's struct inlining like this:

```
// secrets package
type Credential struct {
    AccessKey string `json:"access_key" url:"access-key"`
    SecretKey string `json:"secret_key" url:"secret-key"`
}

// admin package
type UserKeySpec struct {
    secrets.Credential
    User      string `json:"user"`
    // Request fields
    UID         string `url:"uid"`     // The user ID to receive the new key
    SubUser     string `url:"subuser"` // The subuser ID to receive the new key
    KeyType     string `url:"key-type"`
    GenerateKey *bool  `url:"generate-key"` // Generate a new key pair and add to the existing keyring
}

```

Reply:
// behavior:
	//   - overwriteExistingCreds == false:
	//     - creds == <non-empty> 
	//       - if creating a new user, the user should have these creds
	
Why user should have these creds until they have specific source of truth to these.

	//       - if a user already exists, user's existing creds will be used, even if they are different
	//     - creds == <empty> 
	//       - if creating a new user, generate random creds
	//       - if user already exists, user's existing creds will be used
	//   - overwriteExistingCreds == true:
	//     - creds = <empty> - ERROR: creds cannot be empty
We can still create them looking at the K8S secret access key and secret values and updating the cred with those values.
	//     - creds = <non-empty> - always overwrite existing Creds with input creds

Reply:
Updated

Reply:
Because the user (or CI test environments) may create the configmap before Rook is installed to pre-specify the credentials without wanting to specifically set the annotation.

Reply:
In the documentation we have documented a proper section if you want to create a secret before rook is installed or update the secret later, we should do the following steps.

So that covers the above step.

So I believe users should always add an annotation if the secrets are managed by users, as it being the source of truth.

Reply:
That is a reasonable stance to take, I think. :)